examples/platforms/posix/sim/otns.c

```diff
@@ -0,0 +1,66 @@
+/*
+ *  Copyright (c) 2019, The OpenThread Authors.
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions are met:
+ *  1. Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *  2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *  3. Neither the name of the copyright holder nor the
+ *     names of its contributors may be used to endorse or promote products
+ *     derived from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "platform-posix.h"
+
+#if OPENTHREAD_CONFIG_OTNS_ENABLE
+
+#include <stdbool.h>
+#include <stdio.h>
+#include <string.h>
+
+#include <openthread/platform/alarm-micro.h>
+#include <openthread/platform/alarm-milli.h>
+#include <openthread/platform/diag.h>
+
+#if OPENTHREAD_POSIX_VIRTUAL_TIME == 0
+#error "OTNS does not work with VIRTUAL_TIME=0"
+#endif
+
+void otPlatOtnsStatusPush(const char *fmt, ...)
+{
+    int          n;
+    char         maxline[OT_EVENT_DATA_MAX_SIZE];
+    struct Event event;
+
+    va_list ap;
+    va_start(ap, fmt);
+
+    n = vsnprintf(maxline, sizeof(maxline), fmt, ap);
+
+    va_end(ap);
+
+    event.mDelay      = 0;
+    event.mEvent      = OT_SIM_EVENT_STATUS_PUSH;
+    event.mDataLength = (uint16_t)n;
```

Comment:
By `fail`, do you mean the formatted string exceed capacity of `maxline`? When will `vsnprintf` return `n < 0`?

Reply:
No. I am not sure when `vsnprintf` will fail. Maybe by providing invalid format string. But the function definition says it may fail, then it may fail. I believe it will not fail if you are providing with correct arguments. So, an assertion here is acceptable to me.